### ISSUE TITLE
urllib3.exceptions.ReadTimeoutError: AWSHTTPConnectionPool when connecting to moto server

### ISSUE THREAD

**Issue by MamedovaMaria** [spulec/moto#5329]
(1) I am running a standalone moto_server on a localhost on port 5000 with:
- moto version 3.1.16
- boto3 version 1.24.36
- botocore version 1.27.36

I am trying to create a bucket via:
```
  server_client = boto3.client("s3", endpoint_url="http://127.0.0.1:5000", aws_access_key_id="test", aws_secret_access_key="test")
  server_client.create_bucket(Bucket="my-test-bucket")
```

(2) I am getting a socket timeout exception. 

```
Traceback (most recent call last):
  File "/lib/python3.9/site-packages/urllib3/connectionpool.py", line 426, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/lib/python3.9/site-packages/urllib3/connectionpool.py", line 421, in _make_request
    httplib_response = conn.getresponse()
  File "/..pyenv/versions/3.9.6/lib/python3.9/http/client.py", line 1349, in getresponse
    response.begin()
  File "/..pyenv/versions/3.9.6/lib/python3.9/http/client.py", line 316, in begin
    version, status, reason = self._read_status()
  File "/..pyenv/versions/3.9.6/lib/python3.9/http/client.py", line 277, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/..pyenv/versions/3.9.6/lib/python3.9/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lib/python3.9/site-packages/botocore/httpsession.py", line 253, in send
    urllib_response = conn.urlopen(
  File "/lib/python3.9/site-packages/urllib3/connectionpool.py", line 726, in urlopen
tearDown
    retries = retries.increment(
  File "/lib/python3.9/site-packages/urllib3/util/retry.py", line 386, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/lib/python3.9/site-packages/urllib3/packages/six.py", line 735, in reraise
    raise value
  File "/lib/python3.9/site-packages/urllib3/connectionpool.py", line 670, in urlopen
    httplib_response = self._make_request(
  File "/lib/python3.9/site-packages/urllib3/connectionpool.py", line 428, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/lib/python3.9/site-packages/urllib3/connectionpool.py", line 335, in _raise_timeout
    raise ReadTimeoutError(
urllib3.exceptions.ReadTimeoutError: AWSHTTPConnectionPool(host='127.0.0.1', port=5000): Read timed out. (read timeout=60)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/xx/moto_test.py", line 28, in test_load_data_using_decorators
    server_client.create_bucket(Bucket="my-test-bucket")
  File "/lib/python3.9/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/lib/python3.9/site-packages/botocore/client.py", line 647, in _make_api_call
    http, parsed_response = self._make_request(
  File "/lib/python3.9/site-packages/botocore/client.py", line 667, in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
  File "/lib/python3.9/site-packages/botocore/endpoint.py", line 102, in make_request
    return self._send_request(request_dict, operation_model)
  File "/lib/python3.9/site-packages/botocore/endpoint.py", line 136, in _send_request
    while self._needs_retry(attempts, operation_model, request_dict,
  File "/lib/python3.9/site-packages/botocore/endpoint.py", line 228, in _needs_retry
    responses = self._event_emitter.emit(
  File "/lib/python3.9/site-packages/botocore/hooks.py", line 356, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
  File "/lib/python3.9/site-packages/botocore/hooks.py", line 228, in emit
    return self._emit(event_name, kwargs)
  File "/lib/python3.9/site-packages/botocore/hooks.py", line 211, in _emit
    response = handler(**kwargs)
  File "/lib/python3.9/site-packages/botocore/retryhandler.py", line 183, in __call__
    if self._checker(attempts, response, caught_exception):
  File "/lib/python3.9/site-packages/botocore/retryhandler.py", line 250, in __call__
    should_retry = self._should_retry(attempt_number, response,
  File "/lib/python3.9/site-packages/botocore/retryhandler.py", line 277, in _should_retry
    return self._checker(attempt_number, response, caught_exception)
  File "/lib/python3.9/site-packages/botocore/retryhandler.py", line 316, in __call__
    checker_response = checker(attempt_number, response,
  File "/lib/python3.9/site-packages/botocore/retryhandler.py", line 222, in __call__
    return self._check_caught_exception(
  File "/lib/python3.9/site-packages/botocore/retryhandler.py", line 359, in _check_caught_exception
    raise caught_exception
  File "/lib/python3.9/site-packages/botocore/endpoint.py", line 200, in _do_get_response
    http_response = self._send(request)
  File "/lib/python3.9/site-packages/botocore/endpoint.py", line 244, in _send
    return self.http_session.send(request)
  File "/lib/python3.9/site-packages/botocore/httpsession.py", line 288, in send
    raise ReadTimeoutError(endpoint_url=request.url, error=e)
botocore.exceptions.ReadTimeoutError: Read timeout on endpoint URL: "http://127.0.0.1:5000/my-test-bucket"
```

(3) Additionally, the rest api is not accessible via http://127.0.0.1:5000/moto-api 

```
 root@xxx:~# moto_server -H 0.0.0.0 &
 * Running on all addresses (0.0.0.0)
   WARNING: This is a development server. Do not use it in a production deployment.
 * Running on http://127.0.0.1:5000
127.0.0.1 - - [25/Jul/2022 08:14:43] "GET /moto-api HTTP/1.1" 404 -
```

```
root@xxx:~# curl http://127.0.0.1:5000/moto-api
127.0.0.1 - - [25/Jul/2022 08:20:08] "GET /moto-api HTTP/1.1" 404 -
<!doctype html>
<html lang=en>
<title>404 Not Found</title>
<h1>Not Found</h1>
<p>The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.</p>
```

**Comment by bblommers** [comment]
Hi @MamedovaMaria, which version of `werkzeug` do you have installed? 
If it's the latest (`2.2.x`), try downgrading it to `2.1.2`.

**Comment by NFarrington** [comment]
I've been having a similar issue and can confirm downgrading `werkzeug` from `2.2.x` to `2.1.x` has fixed this for me.

**Comment by MamedovaMaria** [comment]
Thanks a lot guys! 
I installed werkzeug==2.1.2 and it fixed the problem

**Comment by bblommers** [comment]
Thanks for letting us know @NFarrington and @MamedovaMaria! 
Marking it as a bug to fix whatever is wrong on our side that causes these timeouts.

**Comment by Tenzer** [comment]
I think this might be a bug in Werkzeug. I managed to drill it down to be a problem inside this while loop which is used for parsing the URL routing rules: https://github.com/pallets/werkzeug/blob/4c7d9908d199d32a360e88d36544c0c01b058b99/src/werkzeug/routing/rules.py#L589-L637.

When it is passed [the first URL from the list of S3 URLs](https://github.com/spulec/moto/blob/ab045f5bebff4de84ac33053ef764efa32405829/moto/s3/urls.py#L16-L21) which takes a parameter, the while loop keeps running indefinitely with 1 as the value for the `pos` variable.

There's no mention in the release notes from Werkzeug that URLs need to be structured in a different way so I would think the new router in Werkzeug 2.2.0 has a bug around this. I can't find any bug reported that seems to match this though.

(I have only tested with S3)
